proposals/extensions_v2.md

```diff
@@ -0,0 +1,358 @@
+# Modern Extensions
+
+* [x] Proposed
+* [ ] Prototype: Not Started
+* [ ] Implementation: Not Started
+* [ ] Specification: Not Started
+
+Many thanks to those who helped with this proposal.  Esp. @jnm2!
+
+## Summary
+[summary]: #summary
+
+Modern Extensions introduce a new syntax to produce "extension members", greatly expanding on the set of supported members including properties, static methods, operators and constructors (and more), in a clean and cohesive fashion.
+
+This new form subsumes C# 3's "classic extension methods", allowing migration to the new modern form in a semantically *identical* fashion (both at a source and ABI level).
+
+A rough strawman of the syntax is as follows.  In all cases, the extended type is shown to be generic, to indicate handling that complex case:
+
+Note: the strawman is not intended to be final form.  That said, it is useful to see any proposed form with all members to ensure that it's generally comprehensible.  A form that is only good for properties, but not for other members, it likely not appropriate.
+
+```c#
+extension E
+{
+    // Instance method form, replaces `public static int M<X>(this SomeType<X> t, ...) { } 
+    public int M<X>(...) for SomeType<X> t { }
+
+    // Property form:
+    // Auto-properties, or usage of 'field' not allowed as extensions do not have instance state.
+    public int Count<X> for SomeType<X> { get { ... } }
+
+    // Event form:
+    // Note: would have to be the add/remove form.
+    // field-backed events would not be possible as extensions do not have instance state.
+    public event Action E<X> for SomeType<X> { add { } remove { } }
+
+    // Indexer form:
+    public int this<T>[int index] for SomeType<X> { get { ... } }
+
+    // Operator form:
+    public static SomeTypeX<X> operator+ <X>(SomeType<X> s1, SomeType<X> s2) for SomeType<X> { ... }
+
+    // Constructor form:
+    public SomeType<X>() for SomeType<X>
+
+    // *Static* extension method (not possible today).  Called as `Assert.Boo("", "")`
+    public static bool Boo(string s1, string s2) for Assert { }
+}
+```
+
+Without specifying the full grammar changes, the intuition is that we are making the following changes:
+
+```g4
+for-clause
+    | 'for' parameter
+    ;
+
+parameter (no change)
+    | attributes? modifiers? type identifier ...
+    ;
+
+extension
+    | attributes? modifiers? 'extension' identifier { member_declaration* }
+    ;
+
+// For method/property/indexer/operator/constructor/event declarations
+// we are augmenting its syntax to allow type-parameters
+// (if not already allowed) and a for-clause. For example:
+property-declaration
+    | attributes? modifiers identifier type-parameters for-clause property-body;
+```
```

Comment:
```suggestion
```

The use of `parameter` means all of the following are legal, with the same semantic extension methods have today:

```c#
for ref Span<T>
for ref readonly Span<T>
for in Span<T>
for scoped ref Span<T>
for scoped Span<T>
```
```